src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -458,6 +456,27 @@ protected virtual void CompleteTransaction(bool isCommitted)
 				{
 					// Allow transaction completed actions to run while others stay blocked.
 					_bypassLock.Value = true;
+					// Ensure no other session processing is still ongoing. In case of a transaction timeout, the transaction is
+					// cancelled on a new thread even for non-distributed scopes. So, the session could be doing some processing,
+					// and will not be interrupted until attempting some usage of the connection. See #3355.
+					// Thread safety of a concurrent session BeginProcess is ensured by the Wait performed by BeginProcess.
+					var isProcessing = _session.IsProcessing();
```

Comment:
Calling BeginProcess was inadequate: it does not support the session being already closed, it attempt to join the current system transaction if any, ... We actually only need to check if the session is still ongoing some processing.